File: AVL tree/python3/erfan/avl.py
from functools import total_ordering
from collections import deque

@total_ordering
class ext_node():
    parent = None
    height = 0
    right = None
    left = None
    def __init__(self, par):
        self.parent = par

    def __eq__(self, other):
        if type(other) == type(None):
            return True
        return False

    def __ne__(self, other):
        return not self.__eq__(other)

    def __gt__(self, other):
        return False

    def __repr__(self):
        return "ext"

@total_ordering
class Node():
    value = None
    parent = None
    height = 1
    def __init__(self, val):
        self.value = val
        self.left = ext_node(self)
        self.right = ext_node(self)
    
    def update_height(self):
        self.height = max(self.right.height, self.left.height) + 1
        return self.height

    def balance_factor(self):
        return self.left.height - self.right.height

    def __eq__(self, other):
        if type(other) == Node:
            return other.value == self.value
        else:
            return other == self.value
    
    def __gt__(self, other):
        if type(other) == Node:
            # return other.value < self.value
            return float(other.value) < float(self.value)
        elif type(other) == ext_node:
            return True
        else:
            # return other < self.value
            return float(other) < float(self.value)

    def __repr__(self):
        return str(self.value)

class Avl():

    def __init__(self):
        self.root = ext_node(None)

    def __str__(self):
        p = deque()
        p.append(self.root)
        while len(p) > 0:
            c = p.popleft()
            print(" "+c.__repr__()+":"+c.parent.__repr__()+" ", end="")
            if c.left != None:
                p.append(c.left)
            if c.right!= None:
                p.append(c.right)
        return (" END")

    def find(self, val, current = None):
        if current is None:
            current = self.root
        suc = None
        while current != None:
            if current == val:
                return current
            elif current < val:
                current = current.right
            elif current > val:
                suc = current
                current = current.left
        return suc

    def remove(self, val):
        current = self.find(val)
        if current == None or current != val:
            return False
        isRoot = False
        if current == self.root:
            isRoot = True
        left_child = False
        if current < current.parent:
            left_child = True
        if current.left == None: # no left for current
            if current.right == None: # current is a leaf
                # if current == self.root:
                #     self.root = self.root.parent
                #     self.root.left = ext_node(self.root)
                #     self.root.right = ext_node(self.root)
                current = current.parent
                if left_child:
                    del current.left
                    current.left = ext_node(current)
                else:
                    del current.right
                    current.right = ext_node(current)
            else: # current has only right
                current = current.parent
                if left_child:
                    current.left = current.left.right
                    current.left.parent = current
                    current = current.left
                else:
                    current.right = current.right.right
                    current.right.parent = current
                    current = current.right
        elif current.right == None: # current has only left
            current = current.parent
            if left_child:
                current.left = current.left.left
                current.left.parent = current
                current = current.left
            else:
                current.right = current.right.left
                current.right.parent = current
                current = current.right
        else: # current has 2 child : replace with suc
            suc = self.find(val,current.right)
            suc_parent = suc.parent
            #correct parent of removed node
            if left_child:
                current.parent.left = suc
            else:
                current.parent.right = suc
            suc.parent = current.parent
            if current is self.root:
                self.root = suc #if removed node  is root, change tree root to suc
            #correct childs
            current.left.parent = suc
            suc.left = current.left
            if current.right != suc:
                current.right.parent = suc
                suc.right = current.right
                suc_parent.left = ext_node(suc_parent)
            #put current on original postion
            del current
            current = suc
        
        print ("removed")
        self.balance(current)
        if isRoot:
            self.root = current 
        return True

    def add(self, val):
        current = self.root
        left_child = False
        while current != None:
            if current == val:
                return False
            elif current < val:
                current = current.right
                left_child = False
            elif current > val:
                current = current.left
                left_child = True
        if current is self.root: #current is root
            self.root = Node(val)
            self.root.parent = current
            self.root.parent.right = self.root
            # self.root.parent.left = self.root
            print ("added")
        else:
            if left_child:
                par = current.parent
                par.left = Node(val)
                par.left.parent = par
                current = par.left
            else:
                par = current.parent
                par.right = Node(val)
                par.right.parent = par
                current = par.right
            print ("added")
            self.balance(current)
        return True

    def balance(self, current):
        while(current != None):
            current.update_height()
            b = current.balance_factor()
            if b > 1: #left
                print ("balancing "+ current.__repr__() )
                if current.left.balance_factor() < 1 : #left right
                    self.left_rotation(current.left)
                # also for left left
                self.right_rotation(current)

            elif b < -1: # right
                print ("balancing "+ current.__repr__() )
                if current.right.balance_factor() > 1 : # right left
                    self.right_rotation(current.right)
                # also for right right
                self.left_rotation(current)

            current = current.parent

    def right_rotation(self, lroot):
        pivot = lroot.left
        lroot.left = pivot.right
        lroot.left.parent = lroot
        pivot.right = lroot
        par = lroot.parent
        lroot.parent = pivot
        pivot.parent = par
        '''
        if par != None: # its not root of tree
            if lroot > par: # so lroot is right child
                par.right = pivot
            else: # so lroot is left child
                par.left = pivot
        else: # change tree root because root has been rotated
            self.root = pivot
        '''
        if lroot > par: # so lroot is right child
            par.right = pivot
        else: # so lroot is left child
            par.left = pivot
        if par == None: # change tree root because root has been rotated
            self.root = pivot

        lroot.update_height()
        pivot.update_height()

    def left_rotation(self, lroot): # same as right_rotation
        pivot = lroot.right
        lroot.right = pivot.left
        lroot.right.parent = lroot
        pivot.left = lroot
        par = lroot.parent
        lroot.parent = pivot
        pivot.parent = par
        '''
        if par != None: # its not root of tree
            if lroot > par: # so lroot is right child
                par.right = pivot
            else: # so lroot is left child
                par.left = pivot
        else: # change tree root because root has been rotated
            self.root = pivot
        '''
        if lroot > par: # so lroot is right child
            par.right = pivot
        else: # so lroot is left child
            par.left = pivot
        if par == None: # change tree root because root has been rotated
            self.root = pivot
        
        lroot.update_height()
        pivot.update_height()

'''
t = Avl()
t.add(1)
t.add(2)
t.add(3)
t.add(4)
t.add(5)
t.add(6)
t.add(7)
t.add(8)
t.add(9)
t.add(10)
t.add(11)
t.add(12)
t.add(13)
t.add(14)
t.remove(1)
t.remove(2)
t.remove(3)
t.remove(4)
t.remove(5)
t.remove(6)
t.remove(7)
t.remove(8)
t.remove(9)
t.remove(10)
t.remove(11)
t.remove(12)
t.remove(13)
t.remove(14)
t.remove(15)
'''
'''
t = Avl()
t.add(3)
t.add(2)
t.add(1)
t.add(0)

print (t)
t.remove(2)
print (t)
''' 
t = Avl()
x = int(input())
for _ in range(x):
    order = input().split()
    if order[0] == "find":
        f = t.find(order[1])
        if f:
            print (f)
        else:
            print ("not found")
    elif order[0] == "add":
        a = t.add(order[1])
        if not a:
            print ("already in there")
    elif order[0] == "remove":
        r = t.remove(order[1])
        if not r:
            print ("does not exist")
    # print (t)

File: avl_tree/Java/Reza/AVL.java

import static java.lang.Double.compare;
import java.util.Scanner;

class Node{
    Node p;   Node l;   Node r;
    int h=0;
    double x;
    Node(double v){
        x=v;
    }
}
class Avl{
    Node r=null;      //root
    int n=0;

public int h(Node u){
    if(u==null)
        return -1;
    return u.h;
}
public int max(int a, int b){
    if(a>=b)
        return a;
    else
        return b;
}
public int bf(Node u){
    if(u==null)
        return 0;
    return h(u.r)-h(u.l);
}

public void rr(Node y){                //right rotate
    Node x=y.l;
    if(y.p!=null){
        if(y.p.r==y)
            y.p.r=x;
        else
            y.p.l=x;
    }
    else{
        r=x;
    }
    x.p=y.p;
    y.p=x;
    if(x.r!=null)
        x.r.p=y;
    y.l=x.r;
    x.r=y;
    y.h=1+max(h(y.l),h(y.r));
    x.h=1+max(h(x.l),h(x.r));
}
public void lr(Node y){              //left rotate
    Node x=y.r;
    if(y.p!=null){
        if(y.p.r==y)
            y.p.r=x;
        else
            y.p.l=x;
    }
    else{
        r=x;
    }
    x.p=y.p;
    y.p=x;
    if(x.l!=null)
        x.l.p=y;
    y.r=x.l;
    x.l=y;
    y.h=1+max(h(y.l),h(y.r));
    x.h=1+max(h(x.l),h(x.r));
}

public void balance(Node z,Node y,Node x){
    if(z.l==y && y.l==x){
        rr(z);
    }
    else if(z.l==y && y.r==x){
        lr(y);
        rr(z);
    }
    else if(z.r==y && y.r==x){
        lr(z);
    }
    else if(z.r==y && y.l==x){
        rr(y);
        lr(z);
    }
}
Node find(double v) {
    Node w = r, z = null;
    while (w != null) {
        int comp = compare(v, w.x);
        if (comp < 0) {
            z=w;
            w = w.l;
        } else if (comp > 0) {
            w = w.r;
        } else {
            return w;
        }
    }
    if(z!=null)
        return z;
    else
    {
        return null;
    }
}

public boolean add(double v){
        Node u=r;
        Node p=null;
        if(r==null){
            r=new Node(v);
            System.out.println("added");
            n++;
        }
        else{
            while(u!=null){
                if(v<u.x){
                    p=u;
                    u=u.l;
                }
                else if(v>u.x){
                    p=u;
                    u=u.r;
                }
                else{
                    System.out.println("already in there");
                    return false;
                }
            }
            u=new Node(v);
            if(v<p.x)
                p.l=u;
            else
                p.r=u;
            u.p=p;
            
            System.out.println("added");
            n++;
            
            Node x=null;
            while(p!=null){
                p.h=1+max(h(p.l),h(p.r));
                if(1<bf(p) || bf(p)<-1 ){
                    System.out.println("balancing"+" "+p.x);
                    balance(p,u,x);
                }
                x=u;
                u=p;    
                p=p.p;
            }
        }  
        
        return true;
}
    
void splice(Node u) {
    Node s, p;
    if (u.l != null) {
        s = u.l;
    }
    else {
        s = u.r;
    }
    
    if (u == r) {
        r = s;
        p = null;
    } 
    else {
        p = u.p;
        if (p.l == u) {
            p.l = s;    
        } 
        else {
            p.r = s;
        }
    }
    if (s != null) {
        s.p = p;
    }
    u=null;
}
    
void remove(double v) {
    Node u=find(v);
    Node p=null;
    if(u!=null && u.x==v){
        if (u.l == null || u.r == null) {
            p=u.p;
            splice(u);
        } else {
            Node w = u.l;
            while (w.r != null)
                w = w.r;
            u.x = w.x;
            p=w.p;
            splice(w);
        }
        System.out.println("removed");
        n--;
        if(n==0)
            r=null;
        else{
            Node x=null;
            Node y=null;
            while(p!=null){
                p.h=1+max(h(p.l),h(p.r));
                if(1<bf(p) || bf(p)<-1 ){
                    System.out.println("balancing"+" "+p.x);
                    if(h(p.l)>h(p.r))
                        y=p.l;
                    else
                        y=p.r;
                    if(h(y.l)>h(y.r))
                        x=y.l;
                    else
                        x=y.r;
                    balance(p,y,x);
                    } 
                    p=p.p;
            }
        }
    }
    else System.out.println("does not exist");
}
    
    public void print(Node u){
       if (u == null) {
           //System.out.println(".");
           return;
       }
       if(u.p!=null){
           if(u.p.r==u)
                System.out.println("           "+u.p.x+"        "+"*"+u.x);
           else
                System.out.println("*"+u.x+"        "+u.p.x+"        ");
       }
       else  System.out.println("           *"+u.x);
       print(u.l);
       print(u.r);
    }
   
}
public class  A{
    public static void main (String [] args){
        Scanner var=new Scanner(System.in);
        Avl a=new Avl();
        int n=var.nextInt();
        for(int i=1;i<=n;i++)
        {
            String s=var.next();
            if(s.equals("add")){
                a.add(var.nextDouble());                  
            }
            else if(s.equals("find")){
                Node p=a.find(var.nextDouble());
                if(p!=null)
                    System.out.println(p.x);
                else
                    System.out.println("not found");
            }
            else if(s.equals("remove")){
                a.remove(var.nextDouble());                  
            }
        }
        /*a.print(a.r);
        System.out.println(a.r.x);*/
    }
}
